### PR TITLE
[ARM] `az resourcemanagement private-link` and `az private-link association`: Add new command groups

### PR DESCRIPTION
**Pull request by tjegbejimba**
**Description**
This PR enables private link commands to create, update, list, show, and delete resource manager private links and private link associations. Microsoft.Resources is updated to use the latest Python SDK. 

**Testing Guide**

**History Notes**

[ARM] `az resourcemanagement private-link create`: Create Resource management private link
[ARM] `az resourcemanagement private-link delete`: Delete Resource management private link
[ARM] `az resourcemanagement private-link show`: Get Resource management private link
[ARM] `az resourcemanagement private-link list`: List Resource management private link
[ARM] `az private-link association create`: Create private link association
[ARM] `az private-link association delete`: Delete private link association
[ARM] `az private-link association show`: Get private link association
[ARM] `az private-link association list`: List private link association

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
private links

File: src/azure-cli/azure/cli/command_modules/resource/commands.py [hunk 1]
```diff
@@ -532,3 +544,15 @@ def load_command_table(self, _):
         g.custom_command('publish', 'publish_bicep_file')
         g.custom_command('version', 'show_bicep_cli_version')
         g.custom_command('list-versions', 'list_bicep_cli_versions')
+
+    with self.command_group('resourcemanagement private-link', resource_resourcemanagementprivatelink_sdk, resource_type=ResourceType.MGMT_RESOURCE_PRIVATELINKS) as g:
+        g.custom_command('create', 'create_resourcemanager_privatelink')
+        g.custom_show_command('show', 'get_resourcemanager_privatelink')
+        g.custom_command('list', 'list_resourcemanager_privatelink')
+        g.custom_command('delete', 'delete_resourcemanager_privatelink')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/commands.py:552]: Do we need to add `confirmation=True` for the delete command?

File: src/azure-cli/azure/cli/command_modules/resource/commands.py [hunk 2]
```diff
@@ -532,3 +544,15 @@ def load_command_table(self, _):
...
+
+    with self.command_group('private-link association', resource_privatelinksassociation_sdk, resource_type=ResourceType.MGMT_RESOURCE_PRIVATELINKS) as g:
+        g.custom_command('create', 'create_private_link_association')
+        g.custom_show_command('show', 'get_private_link_association')
+        g.custom_command('list', 'list_private_link_association')
+        g.custom_command('delete', 'delete_private_link_association')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/commands.py:558]: Same as above

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 3]
```diff
@@ -4106,5 +4106,28 @@ def test_create_template_specs_bicep(self, resource_group, resource_group_locati
         self.cmd('ts delete --template-spec {template_spec_id} --yes')
 
 
+class PrivateLinkTest(ScenarioTest):
+    #@ResourceGroupPreparer(name_prefix='cli_test_template_specs_list', parameter_name='resource_group_one', location='westus')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py:4110]: Why not use `@ResourceGroupPreparer` to create resource groups dynamically?

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 4]
```diff
@@ -4106,5 +4106,28 @@ def test_create_template_specs_bicep(self, resource_group, resource_group_locati
...
+    def test_get_privatelinkassociation(self):
+        self.kwargs.update({
+            'loc': 'centralus',
+            'mg': '24f15700-370c-45bc-86a7-aee1b0c4eb8a',
+            'n': '1d7942d1-288b-48de-8d0f-2d2aa8e03ad4'
+        })
+        self.cmd('private-link association show -m {mg} -n {n}', checks=[
+            self.check('[0].Type', 'Microsoft.Authorization/privateLinkAssociations').
+            self.check('properties.PrivateLink', '{n}')
+        ])
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py:4130]: Could you add tests for the following commands?
- `az resourcemanagement private-link create/list/delete`
- `az private-link association create/list/delete`

**Comment by zhoxing-ms** [comment]
@tjegbejimba Please address these CI issues

**Comment by zhoxing-ms** [comment]
In addition, if you want to write multiple descriptions in the history note, please refer to [submitting-pull-requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests
) and write it in the `History Notes` in PR description, such as: https://github.com/Azure/azure-cli/pull/21843

**Comment by zhoxing-ms** [comment]
Since this PR was submitted late, it cannot catch up with the release of this sprint. We will plan to release it in the next sprint **(05-24)**, is this time acceptable?

**Comment by zhoxing-ms** [comment]
The CI problem you encounter is caused by the breaking change in SDK package `azure-mgmt-resource==21.0.0`, we have draft PR [{ARM} Update customized policy for template by msyyc · Pull Request #22030 · Azure/azure-cli (github.com)](https://github.com/Azure/azure-cli/pull/22030) to add compatible logic for this SDK, please wait for it to merge

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 5]
```diff
@@ -29,31 +30,49 @@ def load_arguments(self, _):
...
-    identity_scope_type = CLIArgumentType(help="Scope that the system assigned identity can access")
-    identity_role_type = CLIArgumentType(options_list=['--role'], help="Role name or id that will be assigned to the managed identity")
+    resource_name_type = CLIArgumentType(
+        options_list=['--name', '-n'], help='The resource name. (Ex: myC)')
+    resource_type_type = CLIArgumentType(
+        help="The resource type (Ex: 'resC'). Can also accept namespace/type format (Ex: 'Microsoft.Provider/resC')")
+    resource_namespace_type = CLIArgumentType(
+        options_list='--namespace', completer=get_providers_completion_list, help="Provider namespace (Ex: 'Microsoft.Provider')")
+    resource_parent_type = CLIArgumentType(required=False, options_list=[
+                                           '--parent'], help="The parent path (Ex: 'resA/myA/resB/myB')")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:45]: Since the default value of `required` is `False`, why we need to define `required=False` repeatedly?

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 6]
```diff
@@ -29,31 +30,49 @@ def load_arguments(self, _):
...
+                                         '--role'], help="Role name or id that will be assigned to the managed identity")
     extended_json_format_type = CLIArgumentType(options_list=['--handle-extended-json-format', '-j'], action='store_true',
                                                 help='Support to handle extended template content including multiline and comments in deployment')
-    deployment_name_type = CLIArgumentType(options_list=['--name', '-n'], required=True, help='The deployment name.')
-    deployment_create_name_type = CLIArgumentType(options_list=['--name', '-n'], required=False, help='The deployment name. Default to template file base name')
-    management_group_id_type = CLIArgumentType(options_list=['--management-group-id', '-m'], required=True, help='The management group id.')
+    deployment_name_type = CLIArgumentType(
+        options_list=['--name', '-n'], required=True, help='The deployment name.')
+    deployment_create_name_type = CLIArgumentType(options_list=[
+                                                  '--name', '-n'], required=False, help='The deployment name. Default to template file base name')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:63]: Same as above

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 7]
```diff
@@ -90,31 +113,42 @@ def load_arguments(self, _):
...
         c.ignore('resource_id')
-        c.argument('resource_name', resource_name_type, arg_group='Resource Id')
-        c.argument('api_version', help='The api version of the resource (omit for the latest stable version)', required=False, arg_group='Resource Id')
-        c.argument('resource_provider_namespace', resource_namespace_type, arg_group='Resource Id')
-        c.argument('resource_type', arg_type=resource_type_type, completer=get_resource_types_completion_list, arg_group='Resource Id')
-        c.argument('parent_resource_path', resource_parent_type, arg_group='Resource Id')
+        c.argument('resource_name', resource_name_type,
+                   arg_group='Resource Id')
+        c.argument('api_version', help='The api version of the resource (omit for the latest stable version)',
+                   required=False, arg_group='Resource Id')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:139]: Same as above

File: src/azure-cli/azure/cli/command_modules/resource/commands.py [hunk 8]
```diff
@@ -215,13 +232,16 @@ def load_command_table(self, _):
...
 
     with self.command_group('group', resource_group_sdk, resource_type=ResourceType.MGMT_RESOURCE_RESOURCES) as g:
-        g.command('delete', 'begin_delete', supports_no_wait=True, confirmation=True)
+        g.command('delete', 'begin_delete',
+                  supports_no_wait=True, confirmation=True)
         g.show_command('show', 'get')
         g.command('exists', 'check_existence')
-        g.custom_command('list', 'list_resource_groups', table_transformer=transform_resource_group_list)
+        g.custom_command('list', 'list_resource_groups',
+                         table_transformer=transform_resource_group_list)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/commands.py:240]: Why did you change these formats? This will lead to too much change in this PR and make others uncertain about what the PR mainly does

**Comment by zhoxing-ms** [comment]
It is recommended not to modify the format of these files, which will lead to too many irrelevant changes in this PR

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 9]
```diff
@@ -663,3 +663,36 @@ def load_arguments(self, _):
 
     with self.argument_context('bicep upgrade') as c:
         c.argument('target_platform', arg_type=bicep_target_platform_type)
+
+    with self.argument_context('resourcemanagement private-link create') as c:
+        c.argument('resource_group', arg_type=resource_group_name_type, help='The name of the resource group.')
+        c.argument('name', options_list=['--name', '-n'], help='The name of the resource management private link.')
+        c.argument('location', options_list=['--location', '-l'], help='the region to create the resource management private link')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:670]:
```suggestion
        c.argument('location', arg_type=get_location_type(self.cli_ctx), validator=get_default_location_from_resource_group, help='the region to create the resource management private link')
```

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 10]
```diff
@@ -3702,3 +3702,41 @@ def show_bicep_cli_version(cmd):
 
 def list_bicep_cli_versions(cmd):
     return get_bicep_available_release_tags()
+
+def create_resourcemanager_privatelink(
+    cmd, resource_group, name, location):
+    rcf = _resource_privatelinks_client_factory(cmd.cli_ctx)
+    return rcf.resource_management_private_link.put(resource_group, name, location)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:3709]: Please note that the third parameter of method `rcf.resource_management_private_link.put` is not `location`, but `ResourceManagementPrivateLinkLocation`.  [code link](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/resources/azure-mgmt-resource/azure/mgmt/resource/privatelinks/v2020_05_01/operations/_operations.py#L632)
Therefore, you need to wrap `location` into `ResourceManagementPrivateLinkLocation` then passing the `ResourceManagementPrivateLinkLocation` into method `rcf.resource_management_private_link.put`
```suggestion
    cmd, resource_group, name, location):
    rcf = _resource_privatelinks_client_factory(cmd.cli_ctx)
    ResourceManagementPrivateLinkLocation = get_sdk(cmd.cli_ctx, ResourceType.MGMT_RESOURCE_PRIVATELINKS,
                                                    'ResourceManagementPrivateLinkLocation', mod='models')
    resource_management_private_link_location = ResourceManagementPrivateLinkLocation(location=location)
    return rcf.resource_management_private_link.put(resource_group, name, resource_management_private_link_location)
```

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:3709]:
```suggestion
    cmd, resource_group, name, location):
    rcf = _resource_privatelinks_client_factory(cmd.cli_ctx)
    ResourceManagementPrivateLinkLocation = get_sdk(cmd.cli_ctx, ResourceType.MGMT_RESOURCE_PRIVATELINKS,
                                                    'ResourceManagementPrivateLinkLocation', mod='models')
    resource_management_private_link_location = ResourceManagementPrivateLinkLocation(location=location)
    return rcf.resource_management_private_link.put(resource_group, name, resource_management_private_link_location)
```

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:3709]: This is the problem that caused the CLI tests to fail

File: src/azure-cli/service_name.json [hunk 11]
```diff
@@ -419,6 +424,11 @@
     "AzureServiceName": "Azure Resource Manager",
     "URL": "https://docs.microsoft.com/azure/azure-resource-manager/management/overview"
   },
+  {
+    "Command": "az resourcemanagement",
+    "AzureServiceName": "Azure Resource Manager",
+    "URL": "https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/create-private-link-access-portal"
+  },
```
Comment (zhoxing-ms) [inline src/azure-cli/service_name.json:431]: > https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/create-private-link-access-portal

Are you sure the document link of `az resourcemanagement` is `create-private-link-access-portal`?

Comment (tjegbejimba) [reply src/azure-cli/service_name.json:431]: Got confirmation from @jennyhunter-msft that 
>https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/create-private-link-access-portal

Is the correct document link to use for now

Comment (zhoxing-ms) [reply src/azure-cli/service_name.json:431]: Out of curiosity, shouldn't the private-link be just a part of resource management? Shouldn't [Azure Resource Manager documentation | Microsoft Docs](https://docs.microsoft.com/en-us/azure/azure-resource-manager/) be a more appropriate doc link for the `az resourcemanagement` command?

Comment (jennyhunter-msft) [reply src/azure-cli/service_name.json:431]: The Private Link work is a subsection of resource management with several specific constraints. As far as helpfulness to the user, the private link specific documentation would be more useful.

Comment (zhoxing-ms) [reply src/azure-cli/service_name.json:431]: OK, thanks~

**Review by zhoxing-ms — APPROVED** [review]
